ludwig/data/dataset/ray.py

```diff
@@ -336,6 +351,11 @@ def _prepare_batch(self, batch: pd.DataFrame) -> Dict[str, np.ndarray]:
             else:
                 res[c] = batch[c].to_numpy()
 
+        if self.augmentation_pipeline:
```

Comment:
Nice! The only issue with this approach is it does the augmentation on the "consumer" process (the one doing the forward / backward pass), when it would be more efficient to offload this to the data ingest processes (the "producers"). Looks like we removed the code that showed and example of this in practice, but I found it by going back in the history a bit:

https://github.com/ludwig-ai/ludwig/blob/bec48803f859b64ca439d6b964dde766c197861c/ludwig/data/dataset/ray.py#L413

Notice the `pipeline.map_batches()` call. What this is doing is lazily defining a set of transformations to perform on the input data before sending it to the local process for consumption.

In this case, we can define a similar transformation for data augmentation like so:

```
    def _augment_batch_fn(self):
        pipeline = self.augmentation_pipeline

        def augment_batch(df: pd.DataFrame) -> pd.DataFrame:
            for c, augmentations in self.augmentation_pipeline.items():
                logger.info(f"RayDatasetBatcher applying augmentation pipeline to batch for feature {c}")
                df[c] = augmentations(torch.tensor(df[c]))
            return df

...

augment_batch = self._augment_batch_fn()

for batch in pipeline.map_batches(augment_batch, batch_size=batch_size, batch_format="pandas").iter_batches( 
    prefetch_blocks=0, batch_size=batch_size, batch_format="pandas"
):
    res = self._prepare_batch(batch)
```

Reply:
@tgaddair   Got it.  I need to move the execution of augmentation from the "consumer" task to the "producer" task.  I'll take a look at the cited code and if I have questions, I'll reach back out.

thank you for the guidance.

Reply:
@tgaddair   Made the change as requested:  moved executing augumentation from "consumer" to "producer" process.

Turns out `_augment_batch_fn()` was bit more involved.  Since `map_batches()` uses `"pandas"` format, the data seen in the function is a `pd.Series`.  I had to extract out the underlying numpy arrays, consolidate into a batch of np.arrays to feed into the augmentation pipeline.  Once this was done, then split the batch into the individual arrays to make it conform to the incoming `pandas` dataframe shape.  

At this point, no run-time exceptions and my simple end-to-end test program with synthetic data runs to completion.

Here is an example of the log message when augmentation is executed.  As you can see the augmentation is occurring id three diffferent processes.
```
(_map_block_nosplit pid=1723) RayDatasetBatcher applying augmentation pipeline to batch for feature my_image_P6DoV_
(_map_block_nosplit pid=1723) RayDatasetBatcher applying augmentation pipeline to batch for feature my_image_P6DoV_
(_map_block_nosplit pid=1720) RayDatasetBatcher applying augmentation pipeline to batch for feature my_image_P6DoV_
(_map_block_nosplit pid=1720) RayDatasetBatcher applying augmentation pipeline to batch for feature my_image_P6DoV_
(_map_block_nosplit pid=1721) RayDatasetBatcher applying augmentation pipeline to batch for feature my_image_P6DoV_
```

I'm attaching the full log file if you want to see all the messages. 
[ray_with_augmentation.txt](https://github.com/ludwig-ai/ludwig/files/10429484/ray_with_augmentation.txt)



Re: the numpy array manipulations described above, there is an alternative way that could avoid the batching and splitting operations.  I decided to do the above to make it consistent with how the Local augmentation operation runs.  If you want me to try this alternative way let me know.